File: CPP/No61.cpp
/**
 * Created by Xiaozhong on 2020/7/22.
 * Copyright (c) 2020/7/22 Xiaozhong. All rights reserved.
 */

#include <iostream>

using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode *rotateRight(ListNode *head, int k) {
        if (head == nullptr || head->next == nullptr) return head;


        ListNode *current = head;
        int n = 1;
        while (current->next != nullptr) {
            current = current->next;
            n++;
        }
        current->next = head;   // 连接成环
        ListNode *new_head = head;
        ListNode *new_tail = head;
        for (int i = 0; i < n - k % n - 1; ++i) {
            new_tail = new_tail->next;
        }
        new_head = new_tail->next;
        new_tail->next = nullptr;
        return new_head;
    }
};

int main() {
    ListNode listNode1(1);
    ListNode listNode2(2);
    ListNode listNode3(3);
    ListNode listNode4(4);
    ListNode listNode5(5);
    listNode1.next = &listNode2;
    listNode2.next = &listNode3;
    listNode3.next = &listNode4;
    listNode4.next = &listNode5;
    listNode5.next = nullptr;
    Solution s;
    ListNode *head = s.rotateRight(&listNode1, 2);
    while (head != nullptr) {
        cout << head->val << " ";
        head = head->next;
    }
}

File: Java/src/No61.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No61 {
    static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode rotateRight(ListNode head, int k) {
        // 首先把链表为空的特殊情况处理掉
        if (head == null) return null;
        // 然后设置快慢指针
        ListNode fast = head, slow = head;
        int size = 0;
        while (fast != null) {
            size++;
            fast = fast.next;
        }
        k %= size;
        fast = head;
        // 用快指针去找到倒数第 k+1 个元素
        for (int i = 0; i < k; ++i) {
            if (fast.next == null) fast = head;
            else fast = fast.next;
        }
        // 用慢指针去定位到新链表的尾部
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }
        // 将原来的链表连成环
        fast.next = head;
        // 找到新链表的头节点
        head = slow.next;
        // 再把环切断
        slow.next = null;
        return head;
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(1), l2 = new ListNode(2), l3 = new ListNode(3), l4 = new ListNode(4), l5 = new ListNode(5);
        l1.next = l2;
        l2.next = l3;
        l3.next = l4;
        l4.next = l5;
        No61 no61 = new No61();
        no61.rotateRight(l1, 2);
    }
}
